sdk/nodejs/resource.ts

```diff
@@ -208,6 +208,12 @@ export interface ResourceOptions {
      * When set to true, protect ensures this resource cannot be deleted.
      */
     protect?: boolean;
+
+    /**
+     * An optional version, corresponding to the version of the provider plugin that should be used when operating on
+     * this resource.
+     */
+    version?: string;
```

Comment:
It's a bit of a shame that we expose this through to the user.  It doesn't seem like we generally want users to have to think about this.  Is there any reasonable way to make this something that provider code-generation passes along without being part of the bag that custom resources accept as options?

It's almost like we want to split these into two interfaces, one for the contract between users and custom resources, and the other between custom resources and the `Resource` base class super-call (and ultimately `registerResource`).  Would that make sense, or is it just going to be cleaner to do it like this?

Cc also @CyrusNajmabadi for thoughts.

Reply:
Relatedly - I assume as written providers will need to pass through any user-provided value for `version`, allowing users to override this if they want.  Are there scenarios where that can make sense?

Reply:
> It's almost like we want to split these into two interfaces, one for the contract between users and custom resources, and the other between custom resources and the Resource base class super-call (and ultimately registerResource). Would that make sense, or is it just going to be cleaner to do it like this?

I think that it's going to be cleaner to do this. Doing the alternative gets us in a position where we now have to pass another argument bag to a resource, representing "internal" properties that aren't exposed for the user.

Furthermore, I don't think it's insane for a user to want to specify a particular version for a resource or set of resources.

> Relatedly - I assume as written providers will need to pass through any user-provided value for version, allowing users to override this if they want. Are there scenarios where that can make sense?

Suppose version v0.18.10 of the "foobaz" provider package has a bug in it. The "foobazx" library depends on v0.18.10 and hasn't been updated yet. A user has downgraded to "foobaz" v0.18.9 *without changing their package.json* and would like to unblock themselves until "foobazx" updates its package reference.

Whether or not this is a valid scenario, I'm not sure, but it is a powerful tool for working around provider issues.

Reply:
That makes sense - lets just keep things simple and keep `version` exposed publicly.  For the doc comment and note on https://pulumi.io/reference/programming-model.html, let's make sure to note that this is an override for the version information that is inferred from the package being used, and should rarely be needed.

Reply:
> I don't think it's insane for a user to want to specify a particular version for a resource or set of resources.

The only thing I worry about is that we now have two sides of this RPC interface which can disagree on the shape of messages that are going to be sent across the channel.  For example, if you pass a different version for the provider binary from the version of the language projection, is everything going to behave sensibly, or will code in the SDKs expect that properties will be returned by the provider binary which are no longer being returned?

I guess if this option exists and you set it, you're already having a bad time and so we might as well provide you the tools you need to get yourself wedged, but it feels like in general if you are setting this value, you're opening yourself up into a bunch of wacky states that we've never expected before and probably don't have a great way to reason about.

Reply:
>  For example, if you pass a different version for the provider binary from the version of the language projection, is everything going to behave sensibly, or will code in the SDKs expect that properties will be returned by the provider binary which are no longer being returned?

The latter will be true - the SDK and the plugins version in lockstep and in general it's going to be "bad" if we mix and match versions of either. It is especially bad when there have been breaking changes between the two.

> I guess if this option exists and you set it, you're already having a bad time and so we might as well provide you the tools you need to get yourself wedged, but it feels like in general if you are setting this value, you're opening yourself up into a bunch of wacky states that we've never expected before and probably don't have a great way to reason about.

In my mind, we can either leave this thing deliberately undocumented and unexported, leave it documented but explain to users that you really don't want to touch this dial without talking to us, or invent another channel for SDKs to pass this information down to the engine. The first two are definitely the easiest and most expedient. I'm fine with any of these options, though I would (selfishly) prefer the first two in the hopes of keeping our model simple.

Reply:
I think either of the first two make sense and trust you on what you feel is better.  If you think having the switch public but documented as “please don’t touch this” as an escape hatch, then great.